### PR TITLE
Feat improve diagnostics

### PR DESCRIPTION
**Pull request by dstallenberg**
Creates a new SynTest diagnostics package containing:
- several new error types
- a Result object type

Applies the new errors throughout the code and removes all "diagnostic.ts" files from packages.

### CODE REVIEW

File: libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts [hunk 1]
```diff
@@ -58,15 +66,24 @@ export abstract class ControlFlowBasedObjectiveFunction<
     this.branchDistanceCalculator = branchDistanceCalculator;
   }
 
-  protected getMatchingFunctionGraph(nodeId: string) {
+  protected getMatchingFunctionGraph(
+    nodeId: string
+  ): undefined | ControlFlowGraph | ContractedControlFlowGraph {
```
Comment (mitchellolsthoorn) [inline libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts:71]: You can remove undefined

File: libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts [hunk 2]
```diff
@@ -58,15 +66,24 @@ export abstract class ControlFlowBasedObjectiveFunction<
...
-      throw new Error(
-        shouldNeverHappen(ControlFlowBasedObjectiveFunction.name)
+    if (functions_.length > 1) {
+      throw new IllegalStateError(
+        "Found multiple functions with the given nodeId"
+      );
+    }
+
+    if (functions_.length === 0) {
+      // TODO or should we return undefined here?
```
Comment (mitchellolsthoorn) [inline libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts:84]: I think this is fine

File: libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts [hunk 3]
```diff
@@ -188,20 +196,26 @@ export abstract class ControlFlowBasedObjectiveFunction<
...
-      throw new TypeError(shouldNeverHappen(BranchObjectiveFunction.name));
+      throw new ImplementationError("Approach level + Branch Distance is NaN", {
+        context: { objectiveId: id, nodeId: closestCoveredNode.id },
+      });
     }
 
     if (branchDistance === 0) {
       // TODO there can still be a crash inside of the if statement giving this result
-      ControlFlowBasedObjectiveFunction.LOGGER.warn("branch distance is zero");
+      BranchObjectiveFunction.LOGGER.warn("branch distance is zero");
```
Comment (mitchellolsthoorn) [inline libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts:218]: Wrong class

File: libraries/search/lib/operators/selection/TournamentSelection.ts [hunk 4]
```diff
@@ -31,8 +31,13 @@ export function tournamentSelection<T extends Encoding>(
   population: T[],
   tournamentSize: number
 ): T {
-  if (tournamentSize < 2)
-    throw new Error(minimumValue("tournament size", 2, tournamentSize));
+  if (tournamentSize < 2) {
+    // TODO Or should it be a result?
```
Comment (mitchellolsthoorn) [inline libraries/search/lib/operators/selection/TournamentSelection.ts:35]: An exception is fine

File: tsconfig-base.json [hunk 5]
```diff
@@ -68,7 +68,7 @@
     //"inlineSources": true /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */,
 
     /* Experimental Options */
-    //"experimentalDecorators": true /* Enables experimental support for ES7 decorators. */,
+    "experimentalDecorators": true /* Enables experimental support for ES7 decorators. */,
```
Comment (mitchellolsthoorn) [inline tsconfig-base.json:71]: This is not needed

File: libraries/cfg/lib/algorithms/edgeContraction.ts [hunk 6]
```diff
@@ -148,38 +156,56 @@ function bfs(
...
   const isEntry = source === controlFlowGraph.entry.id;
   const isSuccessExit = target === controlFlowGraph.successExit.id;
   const isErrorExit = target === controlFlowGraph.errorExit.id;
 
   if (isEntry && (isSuccessExit || isErrorExit)) {
-    throw new Error(cannotMergeEntryAndExit());
+    return failure(new IllegalStateError("Cannot merge entry and exit nodes"));
   }
+
+  return success(undefined as void);
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply libraries/cfg/lib/algorithms/edgeContraction.ts:208]: Removed

File: libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts (hunk 1 above)
Comment (dstallenberg) [reply libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts]: Removed

File: libraries/cfg/lib/algorithms/edgeContraction.ts [hunk 7]
```diff
@@ -188,34 +214,48 @@ function afterGuards(
...
+            mergeNode1: source,
+            mergeNode2: target,
+            removedAmount: newEdges.length - controlFlowGraph.edges.length,
+          },
+        }
       )
     );
   }
+
+  return success(undefined as void);
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply libraries/cfg/lib/algorithms/edgeContraction.ts:248]: Removed

File: libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts (hunk 3 above)
Comment (dstallenberg) [reply libraries/search/lib/objective/ControlFlowBasedObjectiveFunction.ts]: Fixed.

**Review by mitchellolsthoorn — APPROVED** [review]

**Pull request merged**